src/tpm2-tss-engine-ecc.c

```diff
@@ -796,6 +881,99 @@ tpm2tss_ecc_genkey(EC_KEY *key, TPMI_ECC_CURVE curve, const char *password,
     return (r == TSS2_RC_SUCCESS);
 }
 
+/** Customize the pkey_sm2_sign interface
+ *
+ * Customize the pkey_sm2_sign interface. When there is a tpm key extension value, call the ecdsa_ec_key_sign interface to sign.
+ * @retval 1 on success
+ * @retval 0 on failure
+ */
+static int pkey_sm2_sign(EVP_PKEY_CTX *ctx, unsigned char *sig, size_t *siglen,
+                         const unsigned char *tbs, size_t tbslen)
+{
+    EVP_PKEY *pkey = EVP_PKEY_CTX_get0_pkey(ctx);
+    if (pkey != pkey_sm2)
+        return pkey_sm2_sign_default ? pkey_sm2_sign_default(ctx, sig, siglen, tbs, tbslen) : 0;
+
+    EC_KEY *ec = ec_sm2;
+    if (!ec)
+        return 0;
+
+    if (tpm2tss_ecc_getappdata(ec) == NULL) {
+        /* If this is not a TPM2 key, fall through to software functions. */
+        return pkey_sm2_sign_default ? pkey_sm2_sign_default(ctx, sig, siglen, tbs, tbslen) : 0;
+    }
+    else {
+        /* call tpm sig function */
+        const int sig_sz = ECDSA_size(ec);
+        if (sig_sz <= 0) {
+            return 0;
+        }
+        if (sig == NULL) {
+            *siglen = (size_t)sig_sz;
+            return 1;
+        }
+        if (*siglen < (size_t)sig_sz)
+            return 0;
+        ECDSA_SIG *sig_res = ecdsa_ec_key_sign(tbs, tbslen, NULL, NULL, ec);
+        int siglen_temp = i2d_ECDSA_SIG(sig_res, &sig);
+        ECDSA_SIG_free(sig_res);
+        if (siglen_temp < 0)
+            return 0;
+        *siglen = siglen_temp;
+    }
+
+    return 1;
+}
+
+static int sm2_ctrl_wrap(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+{
+    switch (type) {
+    case EVP_PKEY_CTRL_DIGESTINIT:
+        /* Ignore explicit digest selection.*/
+        return 1;
+    default:
+        return sm2_ctrl_default ? sm2_ctrl_default(ctx, type, p1, p2) : 1;
+    }
+}
+
+static int sm2_ctrl_str_wrap(EVP_PKEY_CTX *ctx, const char *type, const char *value)
+{
+    if (type) {
+        if (!strcmp(type, "digest") || !strcmp(type, "md")) {
+            /* Swallow "digest=sm3" coming from -sm3 to avoid unsupported ctrl; always return success. */
+            return 1;
+        }
+        if (!strcmp(type, "distid") || !strcmp(type, "sm2_id")) {
+            /* Let default ctrl_str handle ID parsing/storage. */
+            return sm2_ctrl_default ? sm2_ctrl_default(ctx, EVP_PKEY_CTRL_SET1_ID, (int)strlen(value), (void*)value) : 1;
+        }
+    }
+    return sm2_ctrl_str_default ? sm2_ctrl_str_default(ctx, type, value) : 1;
+}
+
+static int pkey_sm2_digest_custom_wrap(EVP_PKEY_CTX *ctx, EVP_MD_CTX *mctx)
+{
+    int ret;
+    int id_len = 0;
+    unsigned char id[] = SM2_DEFAULT_USERID;
```

Comment:
No need to copy the default userid to the stack, you can pass SM2_DEFAULT_USERID directly to the function.